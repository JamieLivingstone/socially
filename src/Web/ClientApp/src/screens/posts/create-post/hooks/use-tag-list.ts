import axios from 'axios';
import { useState } from 'react';
import { useQuery } from 'react-query';
import { useDebouncedCallback } from 'use-debounce';

export type TagListResponse = {
  hasNextPage: boolean;
  hasPreviousPage: boolean;
  items: Array<{ name: string }>;
  pageNumber: number;
  totalCount: number;
  totalPages: number;
};

export function useTagList() {
  const [filter, setFilter] = useState('');

  const { data, isLoading, isError, error, refetch } = useQuery(
    ['tagList', filter],
    async function fetchTagListRequest() {
      const { data } = await axios.get<TagListResponse>('/api/v1/tags', {
        params: {
          filter,
          pageNumber: 1,
          pageSize: 10,
        },
      });

      return data;
    },
    { enabled: filter.length > 0, suspense: false },
  );

  return {
    setFilter: useDebouncedCallback(setFilter, 300),
    tags: data?.items ?? [],
    isLoading,
    isError,
    error,
    refetch,
  };
}
