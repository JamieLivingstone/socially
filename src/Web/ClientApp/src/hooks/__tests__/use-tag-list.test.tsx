import { act, renderHook } from '@testing-library/react-hooks';
import * as moxios from 'moxios';

import { TagListResponse, useTagList } from '../use-tag-list';
import { wrapper } from './utils';

describe('useTagList', () => {
  beforeEach(() => {
    moxios.install();
  });

  afterEach(() => {
    moxios.uninstall();
  });

  describe('given no search filter', () => {
    test('does not fetch tags', () => {
      const { result } = renderHook(() => useTagList(), { wrapper });

      expect(result.current.isLoading).toEqual(false);
      expect(moxios.requests.count()).toEqual(0);
    });
  });

  describe('given a search filter', () => {
    test('fetches tags', async () => {
      const response: TagListResponse = {
        items: [{ name: 'tag-one' }],
        pageNumber: 1,
        totalPages: 1,
        totalCount: 1,
        hasPreviousPage: false,
        hasNextPage: false,
      };

      moxios.stubRequest(new RegExp('/api/v1/tags.*'), {
        status: 200,
        response,
      });

      const { result, waitForNextUpdate } = renderHook(() => useTagList(), { wrapper });

      act(() => result.current.setFilter('tag-on'));

      await waitForNextUpdate();

      expect(moxios.requests.mostRecent().url).toMatchSnapshot();
      expect(result.current.tags).toMatchSnapshot();
    });
  });
});
