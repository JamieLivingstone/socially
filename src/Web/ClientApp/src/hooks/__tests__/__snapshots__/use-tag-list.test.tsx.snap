// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTagList given a search filter fetches tags 1`] = `"/api/v1/tags?filter=tag-on&pageNumber=1&pageSize=10"`;

exports[`useTagList given a search filter fetches tags 2`] = `
Array [
  Object {
    "name": "tag-one",
  },
]
`;
